course5/homework04/Stream.cc

```diff
@@ -0,0 +1,79 @@
+#include "Stream.h"
+#include <cstdio>
+#include <string>
+#include <stdexcept>
+
+Stream::Stream(const char* file)
+{
+	if ((m_fileptr = std::fopen(file, "w")) == nullptr)
+	{
+		throw (std::runtime_error("Cannot open file\n")); 
+	}
+}
+
+Stream::Stream(FILE* stream)
+{
+	m_fileptr = stream;
+}
+
+Stream::~Stream()
+{
+	if (m_fileptr)
+	{
+		std::fclose(m_fileptr);
+	}
+}
+
+Stream& Stream::operator << (int x)
+{
+	std::fprintf(m_fileptr, "%d",x );
+	return *this; 
+}
+
+Stream& Stream::operator << (long l)
+{
+	std::fprintf(m_fileptr, "%ld", l);
+	return *this; 
+}
+
+Stream& Stream::operator << (const char* s)
+{
+	std::fprintf(m_fileptr, "%s", s);
+	return *this; 
+}
+
+Stream& Stream::operator << (std::string s)
+{
+	std::fprintf(m_fileptr, "%s", s.c_str());
+	if (s == Endl)
```

Comment:
This works as requested, but:
1. User may accidentally trigger a flush, by passing a string of certain value. Violates the Principle of Least Surprise
2. This runtime check (with complexity O(Endl.length())) is needed every time a string is printed. This violates the Don't Pay For What You Are Not Using. The runtime check could be replaced by compile-time dispatch.